### PR TITLE
Improve custom op library handle cleanup

### PR DESCRIPTION
**Pull request by adrianlizarraga**
### Description

- Adds a new C API `OrtApi::RegisterCustomOpsLibrary_V2` that manages the lifetime of dynamic library handles (i.e., calls `dlclose` or `FreeLibrary`).
- Deprecates C API `OrtApi::RegisterCustomOpsLibrary`.
- Adds C++ API wrapper for convenient registering of custom op libraries.
- `PySessionOptions` is now an alias of `OrtSessionOptions`

### Motivation and Context
The current API for registering custom op libraries loads dynamic libraries but requires users to handle the release of the corresponding library handles. Additionally, the user has to make sure to release the library handle _after_ the session has been destroyed (or the program segfaults).

Example of current API usage:
```C++
void* library_handle = nullptr;

{  // Create session in an internal/separate scope so that it is destroyed before dll is released.
  Ort::SessionOptions session_options;
  Ort::ThrowOnError(Ort::GetApi().RegisterCustomOpsLibrary(session_options, "my_op.dll", &library_handle));

  // ...

  Ort::Session session(*ort_env, "model.onnx", session_options);
  // Run inference ...
}

// OS-specific library handle cleanup
#ifdef _WIN32
bool success = ::FreeLibrary(reinterpret_cast<HMODULE>(library_handle));
// ... 
#else
int retval = dlclose(library_handle);
// ...
#endif
```

The new API automatically cleans up the library and allows the user to write more straightforward code.

```C++
Ort::SessionOptions session_options;
session_options.RegisterCustomOpsLibrary("my_op.dll");

// ...

Ort::Session session(*ort_env, "model.onnx", session_options);
// Run inference ...
```

### CODE REVIEW

File: onnxruntime/core/session/abi_session_options.cc [hunk 1]
```diff
@@ -20,6 +20,40 @@ OrtSessionOptions::OrtSessionOptions(const OrtSessionOptions& other)
...
+  if (!library_handle) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Failed to load dynamic library ", library_name);
+  }
+
+  OrtStatus*(ORT_API_CALL * RegisterCustomOps)(OrtSessionOptions * options, const OrtApiBase* api) = nullptr;
+  ORT_RETURN_IF_ERROR(platform_env.GetSymbolFromLibrary(library_handle, "RegisterCustomOps",
+                                                        (void**)&RegisterCustomOps));
+
+  // Call the exported RegisterCustomOps function and store the return value in a unique_ptr.
+  const std::unique_ptr<OrtStatus, decltype(&OrtApis::ReleaseStatus)> status(RegisterCustomOps(this, OrtGetApiBase()),
```
Comment (adrianlizarraga) [inline onnxruntime/core/session/abi_session_options.cc:38]: TODO: Add a test to `test_inference.cc` that exercises a failed called to a custom op library's `RegisterCustomOps` function.

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 2]
```diff
@@ -1270,31 +1282,20 @@ Applies to session load, initialization, etc. Default is 0.)pbdoc")
...
-            OrtSessionOptions s;
-
-            options->custom_op_libraries_.emplace_back(std::make_shared<CustomOpLibrary>(library_path, s));
-
-            // reserve enough memory to hold current contents and the new incoming contents
-            options->custom_op_domains_.reserve(options->custom_op_domains_.size() + s.custom_op_domains_.size());
-            for (size_t i = 0; i < s.custom_op_domains_.size(); ++i) {
-              options->custom_op_domains_.emplace_back(s.custom_op_domains_[i]);
-            }
+            OrtPybindThrowIfError(options->RegisterCustomOpsLibrary(ToPathString(library_name)));
```
Comment (adrianlizarraga) [inline onnxruntime/python/onnxruntime_pybind_state.cc:1295]: Note that the previous implementation of this method passed a default-constructed `OrtSessionOptions` to custom op libraries. Now, we're passing a pointer to the _actual_ `PySessionOptions` to custom op libraries, which is more consistent with a user's expectations (imo).

This change of particular importance to https://github.com/microsoft/onnxruntime/pull/13427 because it allows custom op libraries to extract session configuration entries from the passed in session options. Without this change, a custom op library would only ever see an "empty" session options object.

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 3]
```diff
@@ -1601,11 +1602,12 @@ including arg name, arg type (contains both type and shape).)pbdoc")
...
-          "session_options", [](const PyInferenceSession* sess) -> const PySessionOptions& {
-            const auto& session_options = sess->GetSessionHandle()->GetSessionOptions();
-            return static_cast<const PySessionOptions&>(session_options);
+          "session_options", [](const PyInferenceSession* sess) -> const PySessionOptions* {
+            auto session_options = std::make_unique<PySessionOptions>();
+            session_options->value = sess->GetSessionHandle()->GetSessionOptions();
+            return session_options.release();
           },
-          py::return_value_policy::reference_internal)
+          py::return_value_policy::take_ownership)
```
Comment (adrianlizarraga) [inline onnxruntime/python/onnxruntime_pybind_state.cc:1610]: Note that this is no longer returning a const reference to the inference session's options. Instead, we need to return a heap-allocated `PySessionOptions` that contains a copy of the inference session's options.

`return_value_policy::take_ownership` tells python to manage the lifetime of the returned pointer.

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 4]
```diff
@@ -1152,53 +1098,114 @@ void addObjectMethods(py::module& m, Environment& env, ExecutionProviderRegistra
       sess(m, "SessionOptions", R"pbdoc(Configuration information for a session.)pbdoc");
   sess
       .def(py::init())
-      .def_readwrite("enable_cpu_mem_arena", &PySessionOptions::enable_cpu_mem_arena,
-                     R"pbdoc(Enables the memory arena on CPU. Arena may pre-allocate memory for future usage.
+      .def_property(
+          "enable_cpu_mem_arena",
+          [](const PySessionOptions* options) -> bool { return options->value.enable_cpu_mem_arena; },
+          [](PySessionOptions* options, bool enable_cpu_mem_arena) -> void {
+              options->value.enable_cpu_mem_arena = enable_cpu_mem_arena;
```
Comment (adrianlizarraga) [inline onnxruntime/python/onnxruntime_pybind_state.cc:1105]: Note that `PySessionOptions` is now an alias of `OrtSessionOptions`, which contains a member called `value` of type `onnxruntime::SessionOptions`.

File: onnxruntime/core/common/path_string.h [hunk 5]
```diff
@@ -39,6 +39,10 @@ inline PathChar ToLowerPathChar(PathChar c) {
   return std::towlower(c);
 }
 
+inline std::string PathToUTF8String(const PathString& s) {
+  return ToUTF8String(s);
+}
+
```
Comment (yuslepukhin) [inline onnxruntime/core/common/path_string.h:45]: Why would this be necessary?

File: onnxruntime/core/common/path_string.h [hunk 6]
```diff
@@ -47,6 +51,10 @@ inline PathChar ToLowerPathChar(PathChar c) {
   return std::tolower(c);
 }
 
+inline std::string PathToUTF8String(const PathString& s) {
+  return s;
+}
```
Comment (yuslepukhin) [inline onnxruntime/core/common/path_string.h:56]: This is a NOOP duplicate of line 26.

File: onnxruntime/core/framework/library_handles.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  // Move-only.
+  LibraryHandles(LibraryHandles&& other) noexcept;
+  LibraryHandles& operator=(LibraryHandles&& other) noexcept;
+
+  // Add a library handle that should be unloaded upon destruction of this object.
+  // The `library_name` is used for error reporting if unloading should fail.
+  void Add(PathString library_name, void* library_handle);
+
+ private:
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(LibraryHandles);
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/library_handles.h:38]: >ORT_DISALLOW_COPY_AND [](http://example.com/codeflow?start=2&length=21)

Should be public

File: onnxruntime/core/framework/library_handles.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <utility>
+#include "core/platform/env.h"
+#include "core/common/path_string.h"
+#include "core/common/inlined_containers.h"
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/library_handles.h:11]: >s.h" [](http://example.com/codeflow?start=39&length=4)

`_fwd` would be sufficient in this case and may help avoid issues with CUDA compiles when and if it gets indirectly included into CUDA (wierd things happen).

File: onnxruntime/core/session/abi_session_options.cc [hunk 9]
```diff
@@ -20,6 +20,42 @@ OrtSessionOptions::OrtSessionOptions(const OrtSessionOptions& other)
     : value(other.value), provider_factories(other.provider_factories) {
 }
 
+#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_MINIMAL_BUILD_CUSTOM_OPS)
+onnxruntime::Status OrtSessionOptions::RegisterCustomOpsLibrary(onnxruntime::PathString library_name) {
```
Comment (yuslepukhin) [inline onnxruntime/core/session/abi_session_options.cc:24]: >onnxruntime::PathString [](http://example.com/codeflow?start=64&length=23)

Making a copy here seems redundant.

File: onnxruntime/core/session/abi_session_options_impl.h [hunk 10]
```diff
@@ -18,4 +20,8 @@ struct OrtSessionOptions {
   ~OrtSessionOptions();
   OrtSessionOptions(const OrtSessionOptions& other);
   OrtSessionOptions& operator=(const OrtSessionOptions& other);
+
+#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_MINIMAL_BUILD_CUSTOM_OPS)
+  onnxruntime::Status RegisterCustomOpsLibrary(onnxruntime::PathString library_name);
```
Comment (yuslepukhin) [inline onnxruntime/core/session/abi_session_options_impl.h:25]: >onnxruntime::PathString [](http://example.com/codeflow?start=47&length=23)

const ref may be more appropriate

File: onnxruntime/core/session/abi_session_options_impl.h [hunk 11]
```diff
@@ -6,6 +6,8 @@
 #include <string>
 #include <vector>
 #include <atomic>
+#include "core/common/status.h"
+#include "core/common/path_string.h"
```
Comment (yuslepukhin) [inline onnxruntime/core/session/abi_session_options_impl.h:10]: I am wondering where it gets included since it is an ABI.
Perhaps, forward declarations for Status and PathString would be more appropriate.

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 12]
```diff
@@ -90,8 +90,7 @@ static void TestInference(Ort::Env& env, const std::basic_string<ORTCHAR_T>& mod
                           const std::vector<OutT>& expected_values_y,
                           int provider_type,
                           OrtCustomOpDomain* custom_op_domain_ptr,
-                          const char* custom_op_library_filename,
-                          void** library_handle = nullptr,
+                          const ORTCHAR_T* custom_op_library_filename,
```
Comment (yuslepukhin) [inline onnxruntime/test/shared_lib/test_inference.cc:93]: >ORTCHAR_T [](http://example.com/codeflow?start=32&length=9)

nice

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 13]
```diff
@@ -1220,30 +1218,21 @@ TEST(CApiTest, test_custom_op_library) {
        17, 18, 18, 18, 17,
        17, 17, 17, 17, 17};
 
-  std::string lib_name;
+  onnxruntime::PathString lib_name;
 #if defined(_WIN32)
-  lib_name = "custom_op_library.dll";
+  lib_name = ORT_TSTR("custom_op_library.dll");
```
Comment (yuslepukhin) [inline onnxruntime/test/shared_lib/test_inference.cc:1223]: >("custom_op_library.dll"); [](http://example.com/codeflow?start=21&length=26)

I tried using the name without extension on Windows, it works. I am wondering if it is true on other platforms, then it would simplify the code.

File: onnxruntime/core/session/abi_session_options.cc (hunk 9 above)
Comment (adrianlizarraga) [reply onnxruntime/core/session/abi_session_options.cc:24]: The PathString is std::moved all the way down to the LibraryHandles where it is stored. So, this does not make any unnecessary copies.

If we're going to make a copy of this path string anyway, I find it better to have the function API inform the caller that the passed in string will be copied. If, on the other hand, this function accepted a `const ORTCHAR_T*` then the caller would not necessarily know if this string was going to be copied or if the pointer was just going to be stored in some member. imo

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 14]
```diff
@@ -1601,11 +1602,12 @@ including arg name, arg type (contains both type and shape).)pbdoc")
           },
           py::return_value_policy::reference_internal)
       .def_property_readonly(
-          "session_options", [](const PyInferenceSession* sess) -> const PySessionOptions& {
-            const auto& session_options = sess->GetSessionHandle()->GetSessionOptions();
-            return static_cast<const PySessionOptions&>(session_options);
+          "session_options", [](const PyInferenceSession* sess) -> const PySessionOptions* {
```
Comment (yuslepukhin) [inline onnxruntime/python/onnxruntime_pybind_state.cc:1605]: >const [](http://example.com/codeflow?start=67&length=5)

Suggest not to return const for readability although with pybind it is all inconsequential.

**Comment by pranavsharma** [comment]
The library handle was intentionally not cleaned up by ORT. I don't quite remember the reason. @RyanUnderhill should probably know. Let's understand the reason before adding another API.

**Comment by yuslepukhin** [comment]
> The library handle was intentionally not cleaned up by ORT. I don't quite remember the reason. @RyanUnderhill should probably know. Let's understand the reason before adding another API.

I think it is now taken care of. It is no longer destroyed before the session and the kernels that may use it.

File: onnxruntime/core/session/abi_session_options_impl.h (hunk 11 above)
Comment (adrianlizarraga) [reply onnxruntime/core/session/abi_session_options_impl.h:10]: We discussed in person and decided to leave "path_string.h" as is.

Comment (yuslepukhin) [reply onnxruntime/core/session/abi_session_options_impl.h:10]: We should not rely on indirect includes. This makes things brittle and introduces breakages when the file that includes it decides to remove the include, only to find that other things depend on it. Every header should be self contained.

File: onnxruntime/core/session/abi_session_options.cc (hunk 1 above)
Comment (adrianlizarraga) [reply onnxruntime/core/session/abi_session_options.cc:39]: Done

File: onnxruntime/core/session/abi_session_options_impl.h (hunk 11 above)
Comment (adrianlizarraga) [reply onnxruntime/core/session/abi_session_options_impl.h:10]: Agreed. This is the reason why this file now includes "status.h" explicitly.

File: onnxruntime/core/framework/library_handles.h (hunk 7 above)
Comment (adrianlizarraga) [reply onnxruntime/core/framework/library_handles.h]: Moved to public section.

File: onnxruntime/core/framework/library_handles.h (hunk 8 above)
Comment (adrianlizarraga) [reply onnxruntime/core/framework/library_handles.h]: Updated to use the _fwd version

File: onnxruntime/python/onnxruntime_pybind_state.cc (hunk 14 above)
Comment (adrianlizarraga) [reply onnxruntime/python/onnxruntime_pybind_state.cc]: Updated to non-const return type

File: onnxruntime/test/shared_lib/test_inference.cc (hunk 13 above)
Comment (adrianlizarraga) [reply onnxruntime/test/shared_lib/test_inference.cc:1223]: I looked at the dlopen documentation for linux and wrote a small test program. Unfortunately, linux requires the extension. I'm not sure about other platforms, but it looks like we have to keep this code as is.

File: include/onnxruntime/core/session/onnxruntime_cxx_api.h [hunk 15]
```diff
@@ -478,6 +478,8 @@ struct SessionOptionsImpl : Base<T> {
   SessionOptionsImpl& SetCustomCreateThreadFn(OrtCustomCreateThreadFn ort_custom_create_thread_fn);  ///< Wraps OrtApi::SessionOptionsSetCustomCreateThreadFn
   SessionOptionsImpl& SetCustomThreadCreationOptions(void* ort_custom_thread_creation_options);      ///< Wraps OrtApi::SessionOptionsSetCustomThreadCreationOptions
   SessionOptionsImpl& SetCustomJoinThreadFn(OrtCustomJoinThreadFn ort_custom_join_thread_fn);        ///< Wraps OrtApi::SessionOptionsSetCustomJoinThreadFn
+
+  void RegisterCustomOpsLibrary(const ORTCHAR_T* library_name);  ///< Wraps OrtApi::RegisterCustomOpsLibrary_V2
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_cxx_api.h:482]: Wondering if we should keep the CXX names same as the C names to avoid confusion. I know that the current ver of the API was not even wrapped before this PR.

File: onnxruntime/core/framework/library_handles.cc [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    UnloadLibraries();
+
+    libraries_ = std::move(other.libraries_);
+  }
+
+  return *this;
+}
+
+void LibraryHandles::Add(PathString library_name, void* library_handle) {
+  libraries_.push_back(std::make_pair(std::move(library_name), library_handle));
```
Comment (pranavsharma) [inline onnxruntime/core/framework/library_handles.cc:26]: What if the library_name was already registered before?

Comment (adrianlizarraga) [reply onnxruntime/core/framework/library_handles.cc:26]: The OS-specific library loading utilities for Windows, Linux, and macOS all implement reference counting. Therefore, from the OS's perspective, loading the same library multiple times is fine as long as the loads() are balanced by the same number of unloads(), which this implementation ensures.

From ORT's perspective, loading the same custom ops library multiple times would register duplicate custom op domains, which would result in the following exception.

> error: C++ exception with description "Domain already set in registry"

I think returning that error for duplicate calls to `session_options.RegisterCustomOpsLibrary` is ok. I can add a test that expects this error for duplicate calls.

Adding logic to allow duplicate calls to `session_options.RegisterCustomOpsLibrary` without failure seems complicated. OS-specific search paths may mean that different "library names" can correspond to the same library, and I'm not sure that we can always expect the same library handle to be returned for the same library.

**Comment by pranavsharma** [comment]
LGTM

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Comment by RyanUnderhill** [comment]
> The library handle was intentionally not cleaned up by ORT. I don't quite remember the reason. @RyanUnderhill should probably know. Let's understand the reason before adding another API.

The issue I remember is that 'dlclose' isn't guaranteed to actually do anything, as we discovered on some systems where it wasn't until process exit that the shared library was actually unloaded (and global destructors in it ran).
